src/remote_gym/remote_environment_management.py

```diff
@@ -85,111 +95,239 @@ def start_as_remote_environment(
 
-        tensor_spec_utils.set_bounds(
-            self.action_spec[1], minimum=action_space_bounds[0], maximum=action_space_bounds[1]
-        )
+        tensor_spec_utils.set_bounds(action_spec[1], minimum=action_space_bounds[0], maximum=action_space_bounds[1])
 
         tensor_spec_utils.set_bounds(
-            self.observation_spec[1], minimum=observation_space_bounds[0], maximum=observation_space_bounds[1]
+            observation_spec[1], minimum=observation_space_bounds[0], maximum=observation_space_bounds[1]
         )
 
         tensor_spec_utils.set_bounds(
-            self.observation_spec[2],
-            minimum=self.environment.reward_range[0],
-            maximum=self.environment.reward_range[1],
+            observation_spec[2],
+            minimum=env.reward_range[0],
+            maximum=env.reward_range[1],
         )
 
-        if self.rendering_enabled:
-            assert self.environment.render_mode == "rgb_array", (
+        if enable_rendering:
+            assert env.render_mode == "rgb_array", (
                 "Rendering remote environments is only possible if the `render_mode` attribute "
                 "of the passed environment is 'rgb_array'."
             )
-            self.environment.reset()
-            render_shape = self.environment.render().shape
-            self.observation_spec.update(
+            env.reset()
+            render_shape = env.render().shape
+            observation_spec.update(
                 {3: dm_env_rpc_pb2.TensorSpec(name="rendering", shape=render_shape, dtype=dm_env_rpc_pb2.UINT8)}
             )
             tensor_spec_utils.set_bounds(
-                self.observation_spec[3],
+                observation_spec[3],
                 minimum=0,
                 maximum=255,
             )
 
+        # Pass action and obs layout
+        out_queue.put(action_spec)
+        out_queue.put(observation_spec)
+        initialized = True
+
+        while True:
+            action, reset = in_queue.get()
+            if reset is None:
+                break
+
+            if reset:
+                observation, info = env.reset()
+                reward = 0.0
+                terminated = False
+                truncated = False
+            else:
+                observation, reward, terminated, truncated, info = env.step(action)
+
+            rendering = env.render() if enable_rendering else None
+
+            out_queue.put((observation, reward, terminated, truncated, info, rendering))
+
+        env.close()
+    except Exception as e:
+        if not initialized:
+            out_queue.put(None)
+            out_queue.put(None)
+        out_queue.put(e)
+
+        logging.error("Stacktrace: %s", traceback.format_exc())
+
+
+class ProcessedEnv:
+    def __init__(self, args: dict, enable_rendering: bool, env_id: int):
+        self.in_queue = mp.Queue()
+        self.out_queue = mp.Queue()
+        self.should_reset = True
+
+        self.env_id = env_id
```

Comment:
Is there a use case right now for having an env_id attribute? It seems like most management is done with the `RemoteEnvironmentServicer.environments` data structure.

Reply:
yes, cleaning up again in destroy_environment